src/main/java/com/google/sps/utility/AuthenticationUtility.java

```diff
@@ -91,6 +99,76 @@ public static String generateAuthorizationHeader(HttpServletRequest request) {
     return "Bearer " + authCookie.getValue();
   }
 
+  /**
+   * Creates a Credential object to work with the Google API Java Client. Will handle verifying
+   * userId prior to creating credential. Be sure to do this before creating the credential
+   *
+   * @param request http request from client. Must contain idToken and accessToken
+   * @return a Google credential object that can be used to create an API service instance null if
+   *     userId cannot be verified or accessToken cannot be found
+   */
+  public static Credential getGoogleCredential(HttpServletRequest request) {
+    // Return null if userId cannot be verified
+    try {
+      if (!verifyUserToken(request)) {
+        return null;
+      }
+    } catch (GeneralSecurityException | IOException e) {
+      e.printStackTrace();
+      return null;
+    }
+
+    // Return null if accessToken cannot be found
+    Cookie accessTokenCookie = getCookie(request, "accessToken");
+    if (accessTokenCookie == null) {
+      return null;
+    }
+
+    return getGoogleCredential(accessTokenCookie.getValue());
+  }
+
+  /**
+   * Creates a Credential object to work with the Google API Java Client. Will NOT handle verifying
+   * userId prior to creating credential. Be sure to do this before creating the credential.
+   * Consider using the overloaded method if you need to do a second verification
+   *
+   * @param accessToken String representation of the accessToken to authenticate user
+   * @return a Google credential object that can be used to create an API service instance. null if
+   *     accessToken is empty string
+   */
+  public static Credential getGoogleCredential(String accessToken) {
+    if (accessToken.isEmpty()) {
+      return null;
+    }
+
+    // Build credential object with accessToken
+    Credential.AccessMethod accessMethod = BearerToken.authorizationHeaderAccessMethod();
+    Credential.Builder credentialBuilder = new Credential.Builder(accessMethod);
+    Credential credential = credentialBuilder.build();
+    credential.setAccessToken(accessToken);
+
+    return credential;
+  }
+
+  /**
+   * Get HTTPTransport appropriate for App Engine environment
+   * https://googleapis.dev/java/google-http-client/latest/com/google/api/client/extensions/appengine/http/UrlFetchTransport.html
+   *
+   * @return UrlFetchTransport instance
+   */
+  public static HttpTransport getAppEngineTransport() {
+    return new UrlFetchTransport();
+  }
+
+  /**
+   * Get JsonFactory instance. Use this method to get JsonFactory instances for consistency purposes
+   *
+   * @return JacksonFactory instance
+   */
+  public static JsonFactory getJsonFactory() {
+    return new JacksonFactory();
```

Comment:
Would it make sense to always return the same instance?

Reply:
For all intents and purposes, this should be totally fine. 

We would only need this for building service instances, and Google client APIs don't care which implementation we use

https://developers.google.com/resources/api-libraries/documentation/gmail/v1/java/latest/com/google/api/services/gmail/Gmail.Builder.html

Reply:
Fixed!